src/emu/machine.cpp

```diff
@@ -1116,7 +1116,25 @@ std::string running_machine::nvram_filename(device_t &device) const
 		if (software != nullptr && *software != '\0')
 			result << PATH_SEPARATOR << software;
 
-		std::string tag(device.tag());
+		std::string tag;
+		for (device_t *dev = &device; dev->owner() != nullptr; dev = dev->owner())
+		{
+			std::ostringstream dev_tag;
+			dev_tag << ":" << dev->basetag();
+			device_slot_interface *intf;
+			if (dev->owner() && dev->owner()->interface(intf))
+			{
+				for (romload::system_bios const &bios : romload::entries(dev->rom_region()).get_system_bioses())
+				{
+					if (dev->system_bios() == bios.get_value())
+					{
+						util::stream_format(dev_tag, "_%s", bios.get_name());
+						break;
+					}
+				}
+			}
+			tag = dev_tag.str() + tag;
+		}
```

Comment:
I’m still not entirely sure this always makes sense to do.   For example, should all the NVRAM files for every subdevice get their paths mangled when the system BIOS is changed?  I’d like to solicit a bit more discussion from more people on this aspect of the change.

Reply:
In most cases NVRAM will work (like when it contains RTC data) independent of BIOS version. Only case where it will never work is when nvram data contains flash memory data. Adding that as a flag in nvram interface, and making other flash memory interface which all flash memory based chips will extend will make that possible. Or even separating handling of those two in general. Can work on solution when agreed for best approach.

Reply:
Was thinking more about this and implemented next:
- We will add suffix/postfix only for flash devices ( those have parent rom region name that is equal to device name)
- Since we already have handling if this is going on on driver level, we are not altering those names (since they are anyway in separate directory)